tutorials/synthetic/synthetic_data_generation_example.py

```diff
@@ -0,0 +1,219 @@
+
+
+class BeginsWithLanguageFilter(DocumentFilter):
+    """Filter documents based on language prefix codes.
+
+    Keeps documents that start with any of the specified language codes.
+    Designed to work with prompts that instruct LLMs to prefix responses
+    with language codes (e.g., [EN], [FR], [DE]).
+    """
+
+    def __init__(self, languages: list[str]):
+        self._name = "begins_with_language_filter"
+        self.languages = languages
+
+    def score_document(self, text: str) -> float:
+        if not self.languages:
+            return 1.0  # If no languages are specified, keep all documents
+        return 1.0 if text.startswith(tuple(self.languages)) else 0.0
+
+    def keep_document(self, score: float) -> bool:
+        return score == 1.0
+
+
+def parse_args() -> argparse.Namespace:
+    """Parse command line arguments."""
+    parser = argparse.ArgumentParser(
+        description="Generate synthetic multilingual Q&A data using LLM",
+        formatter_class=argparse.ArgumentDefaultsHelpFormatter,
+    )
+
+    # API Configuration
+    parser.add_argument(
+        "--api-key",
+        type=str,
+        default=os.environ.get("NVIDIA_API_KEY", ""),
+        help="NVIDIA API key (or set NVIDIA_API_KEY environment variable)",
+    )
+    parser.add_argument(
+        "--base-url", type=str, default="https://integrate.api.nvidia.com/v1", help="Base URL for the API endpoint"
+    )
+    parser.add_argument(
+        "--max-concurrent-requests", type=int, default=3, help="Maximum number of concurrent API requests"
+    )
+    parser.add_argument("--max-retries", type=int, default=3, help="Maximum number of retries for failed requests")
+    parser.add_argument("--base-delay", type=float, default=1.0, help="Base delay between retries (in seconds)")
+
+    # Model Configuration
+    parser.add_argument(
+        "--model-name",
+        type=str,
+        default="nvdev/nvidia/llama-3.1-nemotron-70b-instruct",
+        help="Name of the model to use for generation",
+    )
+
+    # Generation Configuration
+    parser.add_argument(
+        "--languages",
+        nargs="+",
+        default=["English", "French", "German", "Spanish", "Italian"],
+        help="Languages to generate Q&A pairs for",
+    )
+    parser.add_argument("--num-samples", type=int, default=100, help="Number of samples to generate")
+    parser.add_argument("--no-filter-languages", action="store_true", help="Do not filter languages")
+    parser.add_argument(
+        "--prompt", type=str, default=None, help="Custom prompt template (must include {language} placeholder)"
+    )
+    parser.add_argument(
+        "--benchmark-output",
+        type=str,
+        default=None,
+        help="Optional file path to save benchmark results",
+    )
+
+    return parser.parse_args()
+
+
+def main() -> None:
+    """Main function to run the synthetic data generation pipeline."""
+    args = parse_args()
```

Comment:
```suggestion
    """Main function to run the synthetic data generation pipeline."""
    client = RayClient()
    client.start()

    args = parse_args()
```

Reply:
It is separate from the executor.

We need to start a Ray cluster via `RayClient` (or bash `ray start ...` is another way) otherwise the example gets stuck for me with:
```
[2025-10-31 14:59:53,888 W 69783 69783] rpc_client.h:153: Failed to connect to GCS at address 127.0.1.1:6379 within 5 seconds.
[2025-10-31 15:00:23,891 W 69783 69783] gcs_client.cc:180: Failed to get cluster ID from GCS server: TimedOut: Timed out while waiting for GCS to become available.
```
printing to the console indefinitely.

Reply:
interesting. It worked well for me tho. Let me try the `RayClient` as well